astropy/stats/histogram.py

```diff
@@ -94,15 +126,11 @@ def calculate_bin_edges(
         else:
             raise ValueError(f"unrecognized bin code: '{bins}'")
 
-        if range:
-            # Check that the upper and lower edges are what was requested.
-            # The current implementation of the bin width estimators does not
-            # guarantee this, it only ensures that data outside the range is
-            # excluded from calculation of the bin widths.
-            if bins[0] != range[0]:
-                bins[0] = range[0]
-            if bins[-1] != range[1]:
-                bins[-1] = range[1]
+        # The current implementation of the bin width estimators does not
+        # guarantee that outer edges are consistent with numpy's approach,
+        # it only ensures that data outside the range is excluded from
+        # calculation of the bin widths.
+        bins[0], bins[-1] = _get_outer_edges(a, range)
```

Comment:
I thought the request in https://github.com/astropy/astropy/issues/8011 was to make the edges consistent with numpy.  When does this function give different results from numpy?

Reply:
The last edge, as reported in #8011, and as output straight from estimators, may be larger than the maximum data point even when no explicit range is given. In contrast, Numpy guarantees that the last edge will be either the maximum data point or the right edge of the user provided range, and never anything else.

Reply:
Should we always use numpy's version?  That could be a breaking change, but we can also consider it to be a bug fix.

Reply:
I'm not addressing the part of #8011 that proposes to delegate computing all edges to numpy here, because I think it'll be a more substantial change, but this is a first step in that direction.